sycl/doc/extensions/proposed/sycl_ext_oneapi_clock.asciidoc

```diff
@@ -0,0 +1,176 @@
+of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 10 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+== Backend support status
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_oneapi_clock`.  The application must check that the device has
+this aspect before submitting a kernel using any of the APIs in this
+extension.  If the application fails to do this, the implementation throws
+a synchronous exception with the `errc::kernel_not_supported` error code
+when the kernel is submitted to the queue.
+
+== Overview
+
+This extension introduces a new free function `clock(scope)`. This function
+allows the user to sample the value from one of three clocks provided by the
+compute units, depending on the value of the scope argument. `scope` is an
+enumeration constant of the new `clock_scope` enum. It should be passed to the
+function to define the clock source; e.g., `clock(clock_scope::sub_group)`
+samples the value from a clock shared by all work-items executing in the same
+sub-group.
+
+This extension also adds a new aspect, `ext_oneapi_clock`, indicating whether
+the feature is supported by the device.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_CLOCK` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the feature-test
+ macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds new device aspect:
+
+```c++
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_clock
+};
+
+} // namespace sycl
+```
+
+[width="100%",%header,cols="50%,50%"]
+|===
+|Aspect
+|Description
+
+|`ext_oneapi_clock`
+|Indicates that the device supports the `sycl::ext::oneapi::experimental::clock()` call.
```

Comment:
Since not all devices are going to support clocks at each scope, I think we need to have some way to query which scopes are supported for the device.

Reply:
We could do this by adding three aspects instead of one.

Reply:
One downside to that is that we cannot use `[[__sycl_detail__::__uses_aspects__()]]` if the aspect requirements are based on runtime arguments. We could if we make the scopes template parameters or overloads, but I don't know if that is to restrictive for common use-cases.

Reply:
Added new aspects in [a01aeb3](https://github.com/intel/llvm/pull/19842/commits/a01aeb3564e27d172dd8d9875d59ddadbcb62d18) and made the function templated.